File: messagebus/src/main/java/com/yahoo/messagebus/routing/RetryTransientErrorsPolicy.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.messagebus.routing;

import com.yahoo.messagebus.ErrorCode;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Implements a retry policy that allows resending of any error that is not fatal. It also does progressive back-off,
 * delaying each attempt by the given time multiplied by the retry attempt.
 *
 * @author Simon Thoresen Hult
 */
public class RetryTransientErrorsPolicy implements RetryPolicy {

    private static final double US = 1000000;
    private final AtomicBoolean enabled = new AtomicBoolean(true);
    private volatile AtomicLong baseDelayUS = new AtomicLong(1000);

    /**
     * Sets whether or not this policy should allow retries or not.
     *
     * @param enabled True to allow retries.
     * @return This, to allow chaining.
     */
    public RetryTransientErrorsPolicy setEnabled(boolean enabled) {
        this.enabled.set(enabled);
        return this;
    }

    /**
     * Sets the base delay in seconds to wait between retries. This amount is multiplied by the retry number.
     *
     * @param baseDelay The time in seconds.
     * @return This, to allow chaining.
     */
    public RetryTransientErrorsPolicy setBaseDelay(double baseDelay) {
        this.baseDelayUS.set((long)(baseDelay*US));
        return this;
    }

    @Override
    public boolean canRetry(int errorCode) {
        return enabled.get() && errorCode < ErrorCode.FATAL_ERROR;
    }

    @Override
    public double getRetryDelay(int retry) {
        long retryMultiplier = 0l;
        if (retry > 1) {
            retryMultiplier = 1L << Math.min(20, retry-1);
        }

        return Math.min(10.0, (retryMultiplier*baseDelayUS.get())/US);
    }
}

File: messagebus/src/vespa/messagebus/routing/retrytransienterrorspolicy.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "retrytransienterrorspolicy.h"
#include <vespa/messagebus/errorcode.h>

namespace mbus {

RetryTransientErrorsPolicy::RetryTransientErrorsPolicy() :
    _enabled(true),
    _baseDelay(0.001)
{}

RetryTransientErrorsPolicy::~RetryTransientErrorsPolicy() = default;

RetryTransientErrorsPolicy &
RetryTransientErrorsPolicy::setEnabled(bool enabled) {
    _enabled = enabled;
    return *this;
}

RetryTransientErrorsPolicy &
RetryTransientErrorsPolicy::setBaseDelay(double baseDelay) {
    _baseDelay = baseDelay;
    return *this;
}

bool
RetryTransientErrorsPolicy::canRetry(uint32_t errorCode) const {
    return _enabled.load(std::memory_order_relaxed) && errorCode < ErrorCode::FATAL_ERROR;
}

double
RetryTransientErrorsPolicy::getRetryDelay(uint32_t retry) const {
    uint64_t retryMultiplier = 0l;
    if (retry > 1) {
        retryMultiplier = 1L << std::min(20u, retry-1);
    }
    return std::min(10.0, _baseDelay.load(std::memory_order_relaxed) * retryMultiplier);
}

}

